mep/people/sitemaps.py

```diff
@@ -0,0 +1,78 @@
+from datetime import date
+
+from django.contrib.sitemaps import Sitemap
+from django.urls import reverse
+from djiffy.models import Canvas
+
+from mep.people.queryset import PersonSolrQuerySet
+
+
+def solr_timestamp_to_date(timestamp):
+    '''Convert solr isoformat date time string to python date.'''
+    # format: 2020-05-12T15:46:20.341Z
+    # django sitemap only includes date, so strip off time
+    yearmonthday = timestamp.split('T')[0]
+    return date(*[int(val) for val in yearmonthday.split('-')])
+
+
+class MemberSitemap(Sitemap):
+    url_route = 'people:member-detail'
+
+    def items(self):
+        return PersonSolrQuerySet().all().only('slug', 'last_modified')
+
+    def location(self, obj):
+        return reverse(self.url_route, kwargs={'slug': obj['slug']})
+
+    def lastmod(self, obj):
+        return solr_timestamp_to_date(obj['last_modified'])
+
+    # NOTE: could set variable priority based on record completeness,
+    # e.g. has card, viaf link, etc.
+
+
+# NOTE: solr index is updated when events change, so member last modified
+# can be used for all associated member pages
+
+
+class MembershipActivitiesSitemap(MemberSitemap):
+    url_route = 'people:membership-activities'
+
+    # NOTE: set low priority for member with no membership activities
+
+
+class BorrowingActivitiesSitemap(MemberSitemap):
+    url_route = 'people:borrowing-activities'
+
+    # NOTE: set low priority for member with no borrowing activities
+
+
+class MemberCardListSitemap(MemberSitemap):
+    url_route = 'people:member-card-list'
+
+    # NOTE: set low priority for member with no cards
+
+
+class MemberCardDetailSitemap(Sitemap):
+    url_route = 'people:member-card-detail'
+
+    def items(self):
+        # find canvas objects associated with members via account cards
+        return Canvas.objects \
+            .filter(manifest__bibliography__account__isnull=False) \
+            .values('short_id',
+                    'manifest__bibliography__account__persons__slug') \
+            .distinct()
+        # NOTE: this does not include the handful of cards that appear
+        # at more than one url, because a member has an event on someone
+        # else's card
+
+    def location(self, obj):
+        return reverse(self.url_route, kwargs={
+            'slug': obj['manifest__bibliography__account__persons__slug'],
+            'short_id': obj['short_id']
+        })
+
+    # Not sure how to calculate last modified; member last indexed
+    # in Solr is probably a decent proxy, but requires a separate query
+    # (and also doesn't account for card image changes)
```

Comment:
i think it's probably not worth it to try to handle card image changes, and _theoretically_ they should be pretty rare...agree that solr last indexed is a decent proxy for last modified here.

Reply:
Any ideas for a good way to get it without querying Solr for each individual member?

Reply:
🤔 wonder if I can query for members with cards and last modified and store it on the sitemap class as a lookup...

Reply:
i really feel like i don't understand this well, but what if we:
- call `super.items()`, which will give us a `SolrQuerySet` with all the last-modified dates
- tweak this `Canvas` queryset to include the member's slug
- somehow join them on the slugs to associate `Canvas`es with dates their member was modified

...does that even make sense??

Reply:
We do have member slug in this queryset, because we need it to generate the urls. 

I like the idea of using `super.items()`, that's slick! I was thinking of converting that query result to a lookup dict by slug and saving on the sitemap object.  Should also filter to just members with cards so we only have ~640 instead of ~5,600.